### PR TITLE
Throttle curve now respects throttle scale and clip

### LINKED ISSUE #3852: Make Throttle Preview in Rates Tab Respect Throttle Limit/Scale Settings
### Is your feature request related to a problem? Please describe

It would be cool if the throttle preview curve respected your throttle limit clip/scale settings. No need to change any functionality, this is purely a cosmetic enhancement to improve understanding of the settings.

### Describe the solution you'd like

scaling the max of the curve if throttle limit is set to scale:
image

showing a flat section at the top of the curve if throttle limit is set to clip
image

### Describe alternatives you've considered

No alternatives relevant.

### Other information

_No response_

### PR DESCRIPTION
**Pull request by ChrisRosser**
I think the maths for this is now sorted so I've opened a fresh PR to keep things clean. Fixes #3852 

@haslinghuis Please feel free to review now. @McGiverGim Please feel free to use this in your own PR and close this PR.

### CODE REVIEW

File: src/js/tabs/pid_tuning.js [hunk 1]
```diff
@@ -1738,29 +1759,51 @@ pid_tuning.initialize = function (callback) {
...
+                    ctrlY = midyr + (midy - midyr) * (canvasWidth - curveClip.x) / (canvasWidth - midx);
+                }
+                context.quadraticCurveTo(ctrlX, ctrlY, curveClip.x, curveClip.y);
+                context.moveTo(curveClip.x, curveClip.y);
+                context.lineTo(canvasWidth, curveClip.y);
+            } else {
             context.quadraticCurveTo(midxl, midyl, midx, midy);
             context.moveTo(midx, midy);
-            context.quadraticCurveTo(midxr, midyr, canvasWidth, 0);
+            context.quadraticCurveTo(midxr, midyr, canvasWidth, topy);
```
Comment (haslinghuis) [inline src/js/tabs/pid_tuning.js:1805]: block indentation :)

File: src/js/tabs/pid_tuning.js [hunk 2]
```diff
@@ -1738,29 +1759,51 @@ pid_tuning.initialize = function (callback) {
...
 
             // math magic by englishman
+            const topy = canvasHeight * (1 - curvescale);
             const midx = canvasWidth * mid;
             const midxl = midx * 0.5;
             const midxr = (((canvasWidth - midx) * 0.5) + midx);
-            const midy = canvasHeight - (midx * (canvasHeight / canvasWidth));
-            const midyl = canvasHeight - ((canvasHeight - midy) * 0.5 *(expo + 1));
-            const midyr = (midy / 2) * (expo + 1);
+            const midy = (canvasHeight - curvescale * (midx * (canvasHeight / canvasWidth)));
```
Comment (haslinghuis) [inline src/js/tabs/pid_tuning.js:1771]: // math magic by englishman :joy: 

- Perhaps camelCase for `topY`, `midXl`, `midYr`, `midX`, `midY`

Comment (ChrisRosser) [reply src/js/tabs/pid_tuning.js]: Agreed! I didn't write that line but I am English...

File: src/js/tabs/pid_tuning.js (hunk 1 above)
Comment (ChrisRosser) [reply src/js/tabs/pid_tuning.js]: Yep!

**Comment by ChrisRosser** [comment]
@haslinghuis I've fixed those comments.

**Comment by ChrisRosser** [comment]
@haslinghuis This is now ready to merge I think! I've added an extra line to make the little blob you get when the receiver is connected move along the line correctly. I've also tested the build.

File: src/js/tabs/pid_tuning.js [hunk 3]
```diff
@@ -1738,29 +1759,52 @@ pid_tuning.initialize = function (callback) {
...
+                    ? getPosfromYBezier(throttleClipY,canvasHeight,midYl, midY)
+                    : getPosfromYBezier(throttleClipY,midY, midYr, topY);
+                let curveClip = getQuadraticCurvePoint(0, canvasHeight, midXl, midYl, midX, midY, clipPos);
+                let ctrlX = curveClip.x / 2;
+                let ctrlY = midYl + (canvasHeight - midYl) * (midX - curveClip.x) / midX;
+                if (throttleLimitPercent > mid){
+                    context.quadraticCurveTo(midXl, midYl, midX, midY);
+                    context.moveTo(midX, midY);
+                    curveClip = getQuadraticCurvePoint(midX, midY, midXr, midYr, canvasWidth, topY, clipPos);
+                    ctrlX = midX + (curveClip.x - midX) /2;
```
Comment (haslinghuis) [inline src/js/tabs/pid_tuning.js:1797]:
```suggestion
                    ctrlX = midX + (curveClip.x - midX) / 2;
```

Comment (haslinghuis) [inline src/js/tabs/pid_tuning.js:1797]:
```suggestion
                    ctrlX = midX + (curveClip.x - midX) / 2;
```

File: src/js/tabs/pid_tuning.js [hunk 4]
```diff
@@ -1738,29 +1759,52 @@ pid_tuning.initialize = function (callback) {
...
+            if (throttleLimitType === THROTTLE_LIMIT_TYPES.CLIP){
+                const throttleClipY = canvasHeight * (1 - throttleLimitPercent);
+                thrpos.y = thrpos.y < throttleClipY ? throttleClipY : thrpos.y;
+                const clipPos = throttleLimitPercent <= mid
+                    ? getPosfromYBezier(throttleClipY,canvasHeight,midYl, midY)
+                    : getPosfromYBezier(throttleClipY,midY, midYr, topY);
+                let curveClip = getQuadraticCurvePoint(0, canvasHeight, midXl, midYl, midX, midY, clipPos);
+                let ctrlX = curveClip.x / 2;
+                let ctrlY = midYl + (canvasHeight - midYl) * (midX - curveClip.x) / midX;
+                if (throttleLimitPercent > mid){
```
Comment (haslinghuis) [inline src/js/tabs/pid_tuning.js:1793]:
```suggestion
                if (throttleLimitPercent > mid) {
```

**Review by haslinghuis — APPROVED** [review]
LGTM!

**Comment by ChrisRosser** [comment]
> LGTM!

Suggestions accepted. Squash and merge!

File: src/js/tabs/pid_tuning.js [hunk 5]
```diff
@@ -1738,29 +1759,52 @@ pid_tuning.initialize = function (callback) {
...
+                    : getQuadraticCurvePoint(midX, midY, midXr, midYr, canvasWidth, topY, (thrPercent - mid) * (1.0 / (1.0 - mid)));
 
             // draw
             context.clearRect(0, 0, canvasWidth, canvasHeight);
             context.beginPath();
             context.moveTo(0, canvasHeight);
-            context.quadraticCurveTo(midxl, midyl, midx, midy);
-            context.moveTo(midx, midy);
-            context.quadraticCurveTo(midxr, midyr, canvasWidth, 0);
+            if (throttleLimitType === THROTTLE_LIMIT_TYPES.CLIP){
```
Comment (haslinghuis) [inline src/js/tabs/pid_tuning.js:1784]:
```suggestion
            if (throttleLimitType === THROTTLE_LIMIT_TYPES.CLIP) {
```

File: src/js/tabs/pid_tuning.js [hunk 6]
```diff
@@ -1715,13 +1715,34 @@ pid_tuning.initialize = function (callback) {
                 };
             }
 
-            /* --- */
+            /*
+            Maths from: https://stackoverflow.com/questions/40918569/quadratic-bezier-curve-calculate-x-for-any-given-y
+            */
+            function getPosfromYBezier(y, startY, cpY, endY){
```
Comment (haslinghuis) [inline src/js/tabs/pid_tuning.js:1721]:
```suggestion
            function getPosfromYBezier(y, startY, cpY, endY) {
```

File: src/js/tabs/pid_tuning.js [hunk 7]
```diff
@@ -1738,29 +1759,52 @@ pid_tuning.initialize = function (callback) {
...
-            const midxr = (((canvasWidth - midx) * 0.5) + midx);
-            const midy = canvasHeight - (midx * (canvasHeight / canvasWidth));
-            const midyl = canvasHeight - ((canvasHeight - midy) * 0.5 *(expo + 1));
-            const midyr = (midy / 2) * (expo + 1);
+            const topY = canvasHeight * (1 - throttleScale);
+            const midX = canvasWidth * mid;
+            const midXl = midX * 0.5;
+            const midXr = (((canvasWidth - midX) * 0.5) + midX);
+            const midY = (canvasHeight - throttleScale * (midX * (canvasHeight / canvasWidth)));
+            const midYl = (canvasHeight - ((canvasHeight - midY) * 0.5 *(expo + 1)));
```
Comment (haslinghuis) [inline src/js/tabs/pid_tuning.js:1772]:
```suggestion
            const midYl = (canvasHeight - ((canvasHeight - midY) * 0.5 * (expo + 1)));
```

File: src/js/tabs/pid_tuning.js [hunk 8]
```diff
@@ -1738,29 +1759,52 @@ pid_tuning.initialize = function (callback) {
             throttleCurve.width = throttleCurve.height *
                 (throttleCurve.clientWidth / throttleCurve.clientHeight);
 
+            const throttleScale = throttleLimitType === THROTTLE_LIMIT_TYPES.SCALE ? throttleLimitPercent : 1;
             const canvasHeight = throttleCurve.height;
             const canvasWidth = throttleCurve.width;
 
             // math magic by englishman
```
Comment (nerdCopter) [inline src/js/tabs/pid_tuning.js:1766]: do we want to keep this comment or append w/ Rosser?

Comment (ChrisRosser) [reply src/js/tabs/pid_tuning.js:1766]: I'm an englishman so I say leave it.  :P

File: src/js/tabs/pid_tuning.js [hunk 9]
```diff
@@ -1715,13 +1715,34 @@ pid_tuning.initialize = function (callback) {
...
+            // y = (1-t)^2 * p0 + 2 * (1-t)*t*p1 + t^2 * p2
+            // y = (p2+p0-2p1)x^2 + 2(p1 - p0)x + p0
+            // 0 = (p2+p0-2p1)x^2 + 2(p1 - p0)x + (p0 - y)
+                const a = startY + endY - 2 * cpY;
+                const b = 2 * (cpY - startY);
+                const c = startY - y;
+                return a == 0 ? -c / b : ( -b - Math.sqrt(Math.pow(b, 2) - 4 * a * c)) / (2 * a);
+            }
 
+            /* --- */
```
Comment (nerdCopter) [inline src/js/tabs/pid_tuning.js:1731]: probably remove empty comment unless something to note here.

**Review by nerdCopter — APPROVED** [review]
- approving. good stuff!
- @haslinghuis , please see my prior, yet insignificant, code-reviews about code-comments before merging.

File: src/js/tabs/pid_tuning.js (hunk 9 above)
Comment (haslinghuis) [inline src/js/tabs/pid_tuning.js:1731]:
```suggestion
```

**Comment by sugaarK** [comment]
if this isn't a bug fix it shouldn't be added to 10.10

**Review by McGiverGim — APPROVED** [review]

**Comment by nerdCopter** [comment]
> if this isn't a bug fix it shouldn't be added to 10.10

I have opinion of the opposite.
- only changes the rendering of the graph.
- does not change behavior when CLIP or SCALE is not selected.
- it "fixes" rendering of CLIP or SCALE. (not rendering properly for a particular setting could arguably be considered a logic-bug.)
- does not require firmware change.

**Comment by sugaarK** [comment]
@nerdCopter  it was mislabeled. If it’s a bug fix then there is no issue

**Pull request merged**